packages/hardhat/contracts/SignatureNFT.sol

```diff
@@ -1,56 +1,98 @@
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 
 contract SignatureNFT is ERC721Tradable {
-    using Counters for Counters.Counter;
-    Counters.Counter public tokenIdCounter;
+
+    error NotAuthorized();
+    error WrongSignature();
 
     // A Kernel address for proper attribution
     address public creator;
 
+    /**
+     * @notice account used to sign the mintings
+     */
+    address public signer;
+
     event NewSignature(address signer, uint256 indexed tokenId, string uri);
 
-    constructor(
-        address _proxyRegistryAddress,
-        address _creator
-    ) ERC721Tradable('Signature Economies', 'SIGN', _proxyRegistryAddress) {
+    modifier onlyCreator() {
+        if (msg.sender != creator) {
+            revert NotAuthorized();
+        }
+        _;
+    }
+
+    constructor(address _proxyRegistryAddress, address _creator, address _signer)
+        ERC721Tradable("Signature Economies", "SIGN", _proxyRegistryAddress)
+    {
         creator = _creator;
+        signer = _signer;
     }
 
     /**
      * @dev Link to Contract metadata https://docs.opensea.io/docs/contract-level-metadata
-    */
+     */
     function contractURI() external pure returns (string memory) {
-        return "https://arweave.net/Ljzre0Drnwhbx8DIoDIMQ0ufC1Ku_HtvRi9IIxNulMQ";
+        return
+            "https://arweave.net/Ljzre0Drnwhbx8DIoDIMQ0ufC1Ku_HtvRi9IIxNulMQ";
     }
 
     /** @notice Set the royalties for the whole contract. Our intention is to set it to 10% in perpetuity.
      *  @param recipient the royalties recipient - will always be pr1s0nart, for regulatory reasons.
      *  @param value royalties value (between 0 and 10000)
-    */
+     */
     function setRoyalties(address recipient, uint256 value)
         external
+        onlyCreator
     {
-        require(msg.sender == creator, "Only the creator of this contract can set and change royalites");
         _setRoyalties(recipient, value);
     }
 
     /**
      * @dev mints a unique NFT from some user-selected piece of text (which we make into an image & store in Arweave)
      * @param uri arweave url
+     * @param tokenId the tokenId to mint
+     * @param signature from `signer` to ensure no race condition on token ids
      * TODO: are there ways to protect this function while still ensuring readers pay gas fees and we don't use onlyOwner?
      */
-    function mintSelected(string memory uri)
-        external
-    {
-        uint256 newTokenId = tokenIdCounter.current();
-        _safeMint(creator, msg.sender, newTokenId);
-        _setTokenURI(newTokenId, uri);
-        tokenIdCounter.increment();
+    function mintSelected(string memory uri, uint256 tokenId, bytes calldata signature) external {
+        // verifies tokenId has been signed / validated by signer
+        _verifySignature(tokenId, uri, signature);
+        
+        // @TODO: verify in eip721 if minting shouldn't come from address(0)
```

Comment:
Good question. The reason I have adapted it is because I am still working off a fork of the NounsDAO contracts and they make the same choice here: https://github.com/nounsDAO/nouns-monorepo/blob/master/packages/nouns-contracts/contracts/base/ERC721.sol#L26